### DISCUSSION TITLE
Why can't the uniform buffer be MAP_WRITE?

### DISCUSSION THREAD

**Discussion by shi-yan** [gpuweb/gpuweb#4293] [Q&A]
when I want to update a uniform buffer, why do I have to create a staging buffer (with COPY_SRC) and copy it to a uniform buffer (UNIFORM | COPY_DST).

why can't the uniform buffer be MAP_WRITE? I hope to understand the underlining mechanism, i.e. is the uniform buffer in local memory?

Also, what the difference between mappedAtCreation == true and mapAsync ? because mappedAtCreation doesn't need MAP_WRITE, but mapAsync does.

**Comment by Kangz** [comment]
Like you guessed, the uniform buffer needs to be in GPU-local memory for the best portable performance. MappedAtCreation is an optimization to avoid extra copies during initialization of a buffer (but yes it is some fake mapping). These twos articles might help get a better idea of what's happening during buffer mapping:

 - https://toji.dev/webgpu-best-practices/buffer-uploads.html
 - https://gpuweb.github.io/gpuweb/explainer/#buffer-mapping
